IdentityCore/src/throttling/MSIDThrottlingService.m

```diff
@@ -0,0 +1,471 @@
+                                                           error:(NSError *_Nullable *_Nullable)error
+{
+    MSIDThrottlingType throttleType = MSIDThrottlingTypeNone;
+    // TODO: Let assume the error here is MSIDError, we will deal with conversion later:
+    // In both scenarios (server error or SSO-ext internal error) broker creates MSALError and return to calling app
+    
+    NSString *httpResponseCode = errorResponse.userInfo[MSIDHTTPResponseCodeKey];
+    NSInteger responseCode = [httpResponseCode intValue];
+    if (responseCode == 429) throttleType = MSIDThrottlingType429;
+    if (responseCode >= 500 && responseCode <= 599) throttleType = MSIDThrottlingType429;
+    NSDate *retryHeaderDate = [self getRetryDateFromErrorResponse:errorResponse];
+    if (retryHeaderDate)
+    {
+        throttleType = MSIDThrottlingType429;
+    }
+    
+    
+    return throttleType;
+}
+
+- (void)updateThrottlingDatabaseWithRequest:(id<MSIDThumbprintCalculatable> _Nonnull)request
+                              errorResponse:(NSError * _Nullable )errorResponse
+                            isSSOExtRequest:(BOOL)isSSOExtRequest
+                                returnError:(NSError *_Nullable *_Nullable)error
+{
+    NSError *localError = nil;
+    MSIDThrottlingType throttleType = [self getThrottleTypeFrom:request
+                                                  errorResponse:errorResponse
+                                                isSSOExtRequest:isSSOExtRequest
+                                                          error:&localError] ;
+    if (localError)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, self.context, BASE_MSG_UPDATING_ERROR, localError);
+        if (error)
+        {
+            *error = localError;
+        }
+        return;
+    }
+    
+    if (throttleType == MSIDThrottlingTypeNone) return;
+    
+    // create throttling record and update to db
+    [self createDBRecordAndUpdateWithRequest:request
+                               errorResponse:errorResponse
+                             isSSOExtRequest:isSSOExtRequest
+                                throttleType:throttleType
+                                 returnError:error];
+    
+    return;
+}
+
+
+/**
+ * If not 429, we check if is appliable for UIRequired:
+ * error response can be: invalid_request, invalid_client, invalid_scope, invalid_grant, unauthorized_client, interaction_required, access_denied
+ */
+- (MSIDThrottlingType)isResponseUIRequiredThrottleType:(NSError *)errorResponse
+{
+    MSIDThrottlingType throttleType = MSIDThrottlingTypeNone;
+    // If not 429, we check if is appliable for UIRequired:
+    // error response can be: invalid_request, invalid_client, invalid_scope, invalid_grant, unauthorized_client, interaction_required, access_denied
+
+    NSSet *uirequiredErrors = [NSSet setWithArray:@[@"invalid_request", @"invalid_client", @"invalid_scope", @"invalid_grant", @"unauthorized_client", @"interaction_required", @"access_denied"]];
```

Comment:
Should we just consider any non-nil OAuth error type as throttlable error?

Reply:
Let discuss it in the other comment above, I think they're same questions.